README.md
# Loriath
Este mod agrega una variedad de items y mecanicas al juego

* [Items](#Items)
	* [Hats](#Hats)
	* [Accesorios](#Accessories)
	* [Mystery Boxes](#mystery-boxes)
* [Tags](#Tags)
* [Enchantments](#Enchantments)

## Items
Se agregan 3 tipos de items

* [Hats](#Hats)
* [Accessories](#Accessories)
* [Mystery Boxes](#mystery-boxes)

### Hats
Estos items son puramente decorativos y se pueden encontrar en cofres y [Mystery Boxes](#mystery-boxes)
Son 121 en total y estan todos bajo la tag `loriath:hats`

### Accessories
Estos items son funcionales, al igual que los [Hats](#Hats) se pueden encontrar en cofres y [Mystery Boxes](#mystery-boxes) son 16 y se encuentran bajo 5 tags diferentes

[comment]: <> (Terminar la tabla)

| Accesorio | Rareza | Descripcion |
| --------- | ------ | ----------- |
|           |        |             |

| Tag                 | Cantidad de items | Items                                                                                           |
| ------------------- | ----------------- | ----------------------------------------------------------------------------------------------- |
| common_accessory    | 3                 | `loriath:adhesive_bandage`, `loriath:bezoar`, `loriath:fast_clock`                              |
| uncommon_accessory  | 3                 | `loriath:cloak_of_invisibility`, `loriath:frog_leg`, `loriath:diving:gear`                      |
| rare_accessory      | 4                 | `loriath:medicated_bandage`, `loriath:shackle`, `loriath:shiny_stone`, `loriath:panic_necklace` |
| epic_accessory      | 3                 | `loriath:lava_charm`, `loriath:digging_claws`, `loriath:frog_flipper`                           |
| legendary_accessory | 3                 | `loriath:destroyer_emblem`, `loriath:hermes_boots`, `loriath:titan_glove`                                                                                                |

### Mystery Boxes
Estos items son muy especiales, ya que al usarlos giras una ruleta que te da un item de una loot table en especifica que depende de la rareza de la misma. Hay 5 rarezas
* Common Mystery Box
* Uncommon Mystery Box
* Rare Mystery Box
* Epic Mystery Box
* Legendary Mystery Box

## Tags
Ademas de las tags de los accesories y hats este mod agrega dos tags mas

| Tag               | Cantidad de items | Description                            |
| ----------------- | ----------------- | -------------------------------------- |
| concrete_powder   | 16                | Todos los bloques de concrete powder   |
| glazed_terracotta | 16                | Todos los bloques de glazed terracotta |

## Enchantments
Se agregan 8 encantamientos, distribuidos de manera que son 7 encantamientos y 1 curse.

| Encantamiento         | Max level | Enchantment Target | Curse | Description                                                                                                                                                                                                                                                               |
| --------------------- | --------- | ------------------ | ----- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| Death Wish            | 1         | Armas              | No    | Tu daño se incrementa en base a la cantidad de vida que te falta. Por otro lado, al tener este encantamiento recibirás 25% más de daño. Ecuación para calcular el daño: `1.0 + (1.0 - entity.getHealth() / entity.getMaxHealth())`                                          |
| Dodge                 | 2         | Pantalones         | No    | Hay una posibilidad de poder evitar cualquier tipo de daño. La posibilidad base es de 12.5% y se multiplica por el nivel del encantamiento                                                                                                                                |
| Elder Guardian Favor  | 1         | Tridente           | No    | Despues de un hit, este encantamiento ataca al enemigo con un láser que no puede ser esquivado o bloqueado (parecido al láser del Elder Guardian). Al estar en el agua, este láser hace un 50% más de daño                                                                |
| Immortality           | 1         | Escudo             | No    | Este encantamiento permite obtener el uso de un tótem al costo de romper el escudo en el cual está aplicado                                                                                                                                                               |
| Phoenix Dive          | 3         | Botas              | No    | Al recibir daño de caída de 3 bloques o más, prende fuego a todas las entidades en un radio de 5 bloques. El daño que hace este fuego disminuye cuanto más lejos esté la entidad de vos. Además te pone unas partículas de fuego re facheras en los pies mientras caminas |
| Smelter               | 1         | Herramientas       | No    | Al romper un bloque, te dará el estado cocinado de ese bloque                                                                                                                                                                                                             |
| Vitality              | 3         | Escudo             | No    | Al tener en la mano un escudo con este encantamiento se te otorgara un contenedor de vida por cada nivel de encantamiento                                                                                                                                                 |
| Incompatibility Curse | 1         | Todo               | Si    | Esta curse no te permite poner ningún otro encantamiento                                                                                                                                                                                                                  |


CHANGELOG.md
## [1.0.8](https://github.com/vicen621/Loriath-Mod/compare/v1.0.7...v1.0.8) (2023-01-17)


### Bug Fixes

* re-enabled code to open mistery boxes ([d46ba30](https://github.com/vicen621/Loriath-Mod/commit/d46ba30887b3cc3d8e09d23081975d096fd9fe1c))

## [1.0.7](https://github.com/vicen621/Loriath-Mod/compare/v1.0.6...v1.0.7) (2023-01-17)


### Bug Fixes

* fixed changelog.md ([b4e480e](https://github.com/vicen621/Loriath-Mod/commit/b4e480e515a5c4ecc1c636cf84ad9ea11b428f78))

## [1.0.6](https://github.com/vicen621/Loriath-Mod/compare/v1.0.5...v1.0.6) (2023-01-17)


### Bug Fixes

* fixed error where the loot table is calling the same event in an infinite loop ([ee25bf8](https://github.com/vicen621/Loriath-Mod/commit/ee25bf89d121d978902da236f63018f4f29bcd97))

## [1.0.5](https://github.com/vicen621/Loriath-Mod/compare/v1.0.4...v1.0.5) (2023-01-17)


### Bug Fixes

* removed last hope ([7bf511b](https://github.com/vicen621/Loriath-Mod/commit/7bf511b705f6873da4458092f79be59fd422ff80))

## [1.0.4](https://github.com/vicen621/Loriath-Mod/compare/v1.0.3...v1.0.4) (2023-01-13)


### Bug Fixes

* modified maricoin_chests.json [no ci] ([3b1c107](https://github.com/vicen621/Loriath-Mod/commit/3b1c107446b90111d313d9fe6675a53d1bc1e592))
* updated node version ([a15f8c8](https://github.com/vicen621/Loriath-Mod/commit/a15f8c8efedc911a4c1796a51001b595a9fcbfe9))

## [1.0.3](https://github.com/vicen621/Loriath-Mod/compare/v1.0.2...v1.0.3) (2023-01-05)


### Bug Fixes

* added AccessoryRender import ([a47550d](https://github.com/vicen621/Loriath-Mod/commit/a47550d9bff25a6a53e6af294e66200190f95f45))
* added HatItem import ([121e181](https://github.com/vicen621/Loriath-Mod/commit/121e181ebc434599a30f72a8d9390e191988ba1e))
* another fix for AccesoryRender ([dd749f8](https://github.com/vicen621/Loriath-Mod/commit/dd749f8fddc640104d9273f08398e3608156893c))
* trying to fix Hat renderer ([e750b94](https://github.com/vicen621/Loriath-Mod/commit/e750b94a93932375da96d061a08b8dda12686515))

## [1.0.2](https://github.com/vicen621/Loriath-Mod/compare/v1.0.1...v1.0.2) (2023-01-05)


### Bug Fixes

* added missing imports ([0276dd8](https://github.com/vicen621/Loriath-Mod/commit/0276dd8166f3af0ba62bdc6fbc1c9a3a5139285b))
* fixed registering render on server ([bb45d8e](https://github.com/vicen621/Loriath-Mod/commit/bb45d8efcb2bf8d5d02c00be1bd598aff4556ff5))

## [1.0.1](https://github.com/vicen621/Loriath-Mod/compare/v1.0.0...v1.0.1) (2023-01-05)


### Bug Fixes

* removed Dash ([9bfc361](https://github.com/vicen621/Loriath-Mod/commit/9bfc361eae2b3f8fc8462780afcb1f606f721218))
* removed dash key in ([e8e8788](https://github.com/vicen621/Loriath-Mod/commit/e8e8788407c2ec10a196db6464f429a2f4144b36))
* removed DashMixin.java [no ci] ([737728f](https://github.com/vicen621/Loriath-Mod/commit/737728f9394035f7a70d5b88faa0bea33f666970))

# 1.0.0 (2023-01-05)


### Bug Fixes

* added cloth-config as dependency ([f111760](https://github.com/vicen621/Loriath-Mod/commit/f111760e3baf12519e96dbdbe913e93e74f65a49))
* added lang keys to en_us.json ([d021852](https://github.com/vicen621/Loriath-Mod/commit/d021852db7fced73eb9196f60dbc490916f507a4))
* fixed accessory typo ([af56f19](https://github.com/vicen621/Loriath-Mod/commit/af56f192a85d760428acce470f1cd847b61428b2))
* fixed branches names ([a6996ac](https://github.com/vicen621/Loriath-Mod/commit/a6996ac5d29378f93c0957d3d30a48863ad61f02))
* fixed version ([46f3b00](https://github.com/vicen621/Loriath-Mod/commit/46f3b009ddc25d5e36aa1f14fb48f8a56b890c79))
* fixed version in build.gradle ([ed08469](https://github.com/vicen621/Loriath-Mod/commit/ed084698b1a155e3c4b3cf1138beb8738b902008))
* merged PlayerEntityMixin's classes ([67a7fce](https://github.com/vicen621/Loriath-Mod/commit/67a7fceee8dd7974e834b2eee1c52de673683581))
* removed dash ([cdef39f](https://github.com/vicen621/Loriath-Mod/commit/cdef39f5dc19974ddff804984799a5477c52dfc7))
* removed dash mixin from mixins ([642105d](https://github.com/vicen621/Loriath-Mod/commit/642105dcc8a34ab0a0041dddf1d31ce7f6f2b932))
* removed DashMixin from server mixins ([f0065b4](https://github.com/vicen621/Loriath-Mod/commit/f0065b4e1b98fd847420c431b0df30fe9993404c))
* removed pocket piston ([74c9f45](https://github.com/vicen621/Loriath-Mod/commit/74c9f45c30a222602cdb2c08d8a9f24f57d5b292))
* removed pocket piston ([9963490](https://github.com/vicen621/Loriath-Mod/commit/99634906a4b554bc68d2b478ec2cceeda9b4241b))
* removed pocket piston ([673afe5](https://github.com/vicen621/Loriath-Mod/commit/673afe5af8fbd8949a0a02926a5ce16cc2d8313a))
* removed unused mixins ([b6fec11](https://github.com/vicen621/Loriath-Mod/commit/b6fec1130260a51879e4b1e3a444d62263160fa2))


### Features

* Added auto-release ([1abe334](https://github.com/vicen621/Loriath-Mod/commit/1abe334b811ba6fc0c9f44c20da589777754e2c2))
* Added Destroyer Emblem Functionality ([2c6d400](https://github.com/vicen621/Loriath-Mod/commit/2c6d4002d587c5e8ae30cd7d5cf794432e6ca347))
* added event LivingDamageEvent ([836a330](https://github.com/vicen621/Loriath-Mod/commit/836a330a30a66f159ecf7fa0a4a079b8deb742f6))
* added Immortality & Vitality enchantments ([ab4b687](https://github.com/vicen621/Loriath-Mod/commit/ab4b687159361d681f4ef22e4ff732f99d459e2a))
* added mistery box particles ([5122832](https://github.com/vicen621/Loriath-Mod/commit/5122832a954b772e2d7ea50ba161029658500cdc))
* added more hats ([2cb2a6f](https://github.com/vicen621/Loriath-Mod/commit/2cb2a6fa6b149646be48962ce9bc719cde8ce822))
* added more particles ([d7be8c4](https://github.com/vicen621/Loriath-Mod/commit/d7be8c425a3276dffca960801f71e59d4b9ac0c3))
* added particles to HermesBoots ([9be988f](https://github.com/vicen621/Loriath-Mod/commit/9be988f1ac5f24d0452b7b0183999ad147904ebb))
* added some more hats ([278ffa6](https://github.com/vicen621/Loriath-Mod/commit/278ffa6665c5121c1274efad6d38b17f146fcb0e))
* added sounds to mistery boxes ([52636e4](https://github.com/vicen621/Loriath-Mod/commit/52636e417e69ed45c77e4d2c3c3b7f6d7b43d021))

src/main/java/io/github/vicen621/loriath/item/ModItems.java
package io.github.vicen621.loriath.item;

import com.google.common.collect.Multimap;
import dev.emi.trinkets.api.SlotReference;
import io.github.vicen621.loriath.LoriathMod;
import io.github.vicen621.loriath.accessories.AccessoryItem;
import io.github.vicen621.loriath.accessories.LavaCharm;
import io.github.vicen621.loriath.accessories.PanicNecklaceItem;
import io.github.vicen621.loriath.accessories.ShinyStone;
import io.github.vicen621.loriath.item.misteryBoxes.MisteryBoxItem;
import io.github.vicen621.loriath.item.misteryBoxes.MisteryBoxRarity;
import net.fabricmc.fabric.api.client.itemgroup.FabricItemGroupBuilder;
import net.fabricmc.fabric.api.item.v1.FabricItemSettings;
import net.minecraft.entity.LivingEntity;
import net.minecraft.entity.attribute.EntityAttribute;
import net.minecraft.entity.attribute.EntityAttributeModifier;
import net.minecraft.entity.attribute.EntityAttributes;
import net.minecraft.entity.effect.StatusEffectInstance;
import net.minecraft.entity.effect.StatusEffects;
import net.minecraft.item.*;
import net.minecraft.nbt.NbtCompound;
import net.minecraft.util.Identifier;
import net.minecraft.util.registry.Registry;

import java.util.UUID;

public class ModItems {

    public static final ItemGroup ITEM_GROUP = FabricItemGroupBuilder.create(
            LoriathMod.id("loriath"))
            .icon(() -> new ItemStack(ModItems.MARICOIN))
            .build();

    public static final Item MARICOIN = registerItem("maricoin", new Item(new FabricItemSettings().group(ITEM_GROUP).maxCount(16)));

    public static final Item COMMON_MISTERY_BOX = new MisteryBoxItem(MisteryBoxRarity.COMMON, new FabricItemSettings().group(ITEM_GROUP).maxCount(1));
    public static final Item UNCOMMON_MISTERY_BOX = new MisteryBoxItem(MisteryBoxRarity.UNCOMMON, new FabricItemSettings().group(ITEM_GROUP).maxCount(1));
    public static final Item RARE_MISTERY_BOX = new MisteryBoxItem(MisteryBoxRarity.RARE, new FabricItemSettings().group(ITEM_GROUP).maxCount(1));
    public static final Item EPIC_MISTERY_BOX = new MisteryBoxItem(MisteryBoxRarity.EPIC, new FabricItemSettings().group(ITEM_GROUP).maxCount(1));
    public static final Item LEGENDARY_MISTERY_BOX = new MisteryBoxItem(MisteryBoxRarity.LEGENDARY, new FabricItemSettings().group(ITEM_GROUP).maxCount(1));

    //public static final Item DASH_SHIELD = registerItem("dash_shield", new FabricShieldItem(new FabricItemSettings().group(ITEM_GROUP).maxDamage(1200), 5, 13, Items.OAK_PLANKS));
    public static final Item FROG_LEG = registerItem("frog_leg", new AccessoryItem() {
        @Override
        public StatusEffectInstance getPermanentEffect() {
            return new StatusEffectInstance(StatusEffects.JUMP_BOOST, 20, 1, true, false);
        }
    });
    public static final Item DIVING_GEAR = registerItem("diving_gear", new AccessoryItem() {

        @Override
        public void onUnequip(ItemStack stack, SlotReference slot, LivingEntity entity) {
            stack.getOrCreateNbt().putInt("divingGearCooldown", 0);
        }

        @Override
        public void tick(ItemStack stack, LivingEntity entity) {
            if (entity.getAir() < 100 && entity.isInsideWaterOrBubbleColumn()) {
                NbtCompound nbt = stack.getOrCreateNbt();
                int cooldown = nbt.getInt("divingGearCooldown");
                if (cooldown <= 0) {
                    entity.setAir(entity.getAir() + 30);
                    nbt.putInt("divingGearCooldown", 60); // 3 seconds
                } else {
                    cooldown--;
                    nbt.putInt("divingGearCooldown", cooldown);
                }
            }
        }
    });
    public static final Item FROG_FLIPPER = registerItem("frog_flipper", new AccessoryItem() {

        @Override
        public void tick(ItemStack stack, LivingEntity entity) {
            if (entity.getAir() < 100 && entity.isInsideWaterOrBubbleColumn()) {
                NbtCompound nbt = stack.getOrCreateNbt();
                int cooldown = nbt.getInt("frogFlipperCooldown");
                if (cooldown <= 0) {
                    entity.setAir(entity.getAir() + 30);
                    nbt.putInt("frogFlipperCooldown", 60); // 3 seconds
                } else {
                    cooldown--;
                    nbt.putInt("frogFlipperCooldown", cooldown);
                }
            }
        }

        @Override
        public StatusEffectInstance getPermanentEffect() {
            return new StatusEffectInstance(StatusEffects.JUMP_BOOST, 20, 1, true, false);
        }

        @Override
        public void onUnequip(ItemStack stack, SlotReference slot, LivingEntity entity) {
            stack.getOrCreateNbt().putInt("frogFlipperCooldown", 0);
        }
    });
    public static final Item HERMES_BOOTS = registerItem("hermes_boots", new AccessoryItem() {
        @Override
        public Multimap<EntityAttribute, EntityAttributeModifier> getModifiers(ItemStack stack,
                                                                               SlotReference slot, LivingEntity entity, UUID uuid) {
            var modifiers = super.getModifiers(stack, slot, entity, uuid);
            modifiers.put(EntityAttributes.GENERIC_MOVEMENT_SPEED, new EntityAttributeModifier(uuid,
                    "loriath:movement_speed", 0.04, EntityAttributeModifier.Operation.ADDITION));
            return modifiers;
        }
    });
    public static final Item BEZOAR = registerItem("bezoar", new AccessoryItem() {
        @Override
        public void tick(ItemStack stack, LivingEntity entity) {
            if (entity.hasStatusEffect(StatusEffects.POISON) && entity.getStatusEffect(StatusEffects.POISON).getAmplifier() == 0) {
                entity.removeStatusEffect(StatusEffects.POISON);
            }
        }
    });
    public static final Item ADHESIVE_BANDAGE = registerItem("adhesive_bandage", new AccessoryItem() {
        @Override
        public void tick(ItemStack stack, LivingEntity entity) {
            if (entity.hasStatusEffect(StatusEffects.WITHER) && entity.getStatusEffect(StatusEffects.WITHER).getAmplifier() == 0) {
                entity.removeStatusEffect(StatusEffects.WITHER);
            }
        }
    });
    public static final Item MEDICATED_BANDAGE = registerItem("medicated_bandage", new AccessoryItem() {
        @Override
        public void tick(ItemStack stack, LivingEntity entity) {
            if (entity.hasStatusEffect(StatusEffects.WITHER) && entity.getStatusEffect(StatusEffects.WITHER).getAmplifier() == 0) {
                entity.removeStatusEffect(StatusEffects.WITHER);
            }
            if (entity.hasStatusEffect(StatusEffects.POISON) && entity.getStatusEffect(StatusEffects.POISON).getAmplifier() == 0) {
                entity.removeStatusEffect(StatusEffects.POISON);
            }
        }
    });
    public static final Item FAST_CLOCK = registerItem("fast_clock", new AccessoryItem() {
        @Override
        public void tick(ItemStack stack, LivingEntity entity) {
            if (entity.hasStatusEffect(StatusEffects.SLOWNESS) && entity.getStatusEffect(StatusEffects.SLOWNESS).getAmplifier() == 0) {
                entity.removeStatusEffect(StatusEffects.SLOWNESS);
            }
        }
    });
    public static final Item SHINY_STONE = registerItem("shiny_stone", new ShinyStone());
    public static final Item LAVA_CHARM = registerItem("lava_charm", new LavaCharm());
    public static final Item TITAN_GLOVE = registerItem("titan_glove", new AccessoryItem() {
        @Override
        public Multimap<EntityAttribute, EntityAttributeModifier> getModifiers(ItemStack stack,
                                                                               SlotReference slot, LivingEntity entity, UUID uuid) {
            var modifiers = super.getModifiers(stack, slot, entity, uuid);
            modifiers.put(EntityAttributes.GENERIC_ATTACK_SPEED, new EntityAttributeModifier(uuid,
                    "loriath:attack_sped", 1, EntityAttributeModifier.Operation.MULTIPLY_TOTAL));
            modifiers.put(EntityAttributes.GENERIC_ATTACK_DAMAGE, new EntityAttributeModifier(uuid,
                    "loriath:attack_damage", 0.1, EntityAttributeModifier.Operation.ADDITION));
            return modifiers;
        }
    });
    public static final Item SHACKLE = registerItem("shackle", new AccessoryItem() {
        @Override
        public Multimap<EntityAttribute, EntityAttributeModifier> getModifiers(ItemStack stack,
                                                                               SlotReference slot, LivingEntity entity, UUID uuid) {
            var modifiers = super.getModifiers(stack, slot, entity, uuid);
            modifiers.put(EntityAttributes.GENERIC_ARMOR, new EntityAttributeModifier(uuid,
                    "loriath:armor", 2, EntityAttributeModifier.Operation.ADDITION));
            return modifiers;
        }
    });
    //TODO
    public static final Item DESTROYER_EMBLEM = registerItem("destroyer_emblem", new AccessoryItem());
    public static final Item PANIC_NECKLACE = registerItem("panic_necklace", new PanicNecklaceItem());

    private static Item registerItem(String name, Item item) {
        return Registry.register(Registry.ITEM, LoriathMod.id(name), item);
    }

    public static void registerModItems() {
        LoriathMod.LOGGER.info("Registering Mod Items for " + LoriathMod.MODID);
    }
}
